resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala

```diff
@@ -0,0 +1,150 @@
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.security.HadoopDelegationTokenManager
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.RpcEndpointRef
+import org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages.UpdateDelegationTokens
+import org.apache.spark.util.ThreadUtils
+
+
+class MesosCredentialRenewer(
+    conf: SparkConf,
+    tokenManager: HadoopDelegationTokenManager,
+    nextRenewal: Long,
+    de: RpcEndpointRef) extends Logging {
+  private val credentialRenewerThread =
+    Executors.newSingleThreadScheduledExecutor(
+      ThreadUtils.namedThreadFactory("Credential Refresh Thread"))
+
+  @volatile private var timeOfNextRenewal = nextRenewal
+
+  private val principal = conf.get("spark.yarn.principal")
+
+  private val (secretFile, mode) = getSecretFile(conf)
+
+  private def getSecretFile(conf: SparkConf): (String, String) = {
+    val keytab64 = conf.get("spark.yarn.keytab", null)
+    val tgt64 = System.getenv("KRB5CCNAME")
+    require(keytab64 != null || tgt64 != null, "keytab or tgt required")
+    require(keytab64 == null || tgt64 == null, "keytab and tgt cannot be used at the same time")
+    val mode = if (keytab64 != null) "keytab" else "tgt"
+    val secretFile = if (keytab64 != null) keytab64 else tgt64
+    logInfo(s"Logging in as $principal with mode $mode to retrieve HDFS delegation tokens")
+    logDebug(s"secretFile is $secretFile")
+    (secretFile, mode)
+  }
+
+  def scheduleTokenRenewal(): Unit = {
+    def scheduleRenewal(runnable: Runnable): Unit = {
+      val remainingTime = timeOfNextRenewal - System.currentTimeMillis()
+      if (remainingTime <= 0) {
+        logInfo("Credentials have expired, creating new ones now.")
+        runnable.run()
+      } else {
+        logInfo(s"Scheduling login from keytab in $remainingTime millis.")
+        credentialRenewerThread.schedule(runnable, remainingTime, TimeUnit.MILLISECONDS)
+      }
+    }
+
+    val credentialRenewerRunnable =
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            val creds = getRenewedDelegationTokens(conf)
+            broadcastDelegationTokens(creds)
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
+              logWarning("Couldn't broadcast tokens, trying again in an hour", e)
+              credentialRenewerThread.schedule(this, 1, TimeUnit.HOURS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    scheduleRenewal(credentialRenewerRunnable)
+  }
+
+  private def getRenewedDelegationTokens(conf: SparkConf): Array[Byte] = {
+    logInfo(s"Attempting to login with ${conf.get("spark.yarn.principal", null)}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174
+    val ugi = if (mode == "keytab") {
```

Comment:
I don't see where it refreshes the delegation tokens until the max-lifetime, then re-login with the keytab to get a new delegation tokens that'll last until the max-lifetime.

Does this skip over the potential issues with expiring delegation tokens (after the max-lifetime, 7 days default) by just re-logging in with the keytab every time the delegation tokens need to refresh, and then grabbing a new set of delegation tokens?

Reply:
Hello @kalvinnchau You are correct, all this does is keep track of when the tokens will expire and renew them at that time. Part of my motivation for doing this is to avoid writing any files to disk (like new TGTs, if that's what you're suggesting). We can simply mount the keytab via the Mesos secrets primitive, then renew the tokens every so often. In order to be consistent I tried to keep this solution as close to YARN as possible.

Reply:
The correct way would be for the credential management code to differentiate between token creation and token renewal; that way it would renew tokens at the renewal internal and create new ones after the max lifetime.

But it seems the original implementation took a shortcut and just creates new one instead of renewing existing ones; changing that would require changes in the credential provider interfaces, so this is enough for now.